* Version 1.2.11
  - Support for 3.2 basecaller and chemistry
  - GFF bug fixes
  - Misc fixes and enhancements

* Version 1.2.10 (> SMRTanalysis 3.1)
  - Update to pysam 0.9.0
  - Recognize new and prospective future partnumbers for the 3.1 timeframe
  - Misc fixes for speeding up dataset and cmp.h5 access

* Version 1.2.9
  - Rename pulseFeature accessors to "baseFeature...", since they
    reflect features accompanying basecalls as opposed to pulse calls;
    add "pulseFeature..." accessors that actually get the pulse
    features (from an internal-mode BAM)
  - Improve FastaWriter performance with IndexedFastaRecords
  - Improve DataSet chunking, filtration, consolidation

* Version 1.2.8
  - Dataset improvements
  - Dataset CLI moved to pbcoretools
  - Support for context flag filtering
  - Misc fixes

* Version 1.2.7...
  - Improved IPython tab completion
  - Can now use .open on Dataset XML files
  - Update to pysam 0.8.4
  - Miscellaneous bugfixes

* Version 1.2.6
  - Recognize sequencing kit C4v2
  - Support for multipart bas files with the new Sequel holenumber format
  - Add readsByHoleNumber to aln file classes
* Version 1.2.5
- Accessors for SNR from BAM reads; needed for Arrow access

* Version 1.2.4
- add GFF gather functionality
- support for 3.0.1 BAM format spec

* Version 1.2.3
- Updated for newer dataset schema
- BAM access speed improvements
- BGZF fixed for large bam.pbi files
- readsInRange iteration order parity between different dataset sources

* Version 1.2.2
- Bug fixes and speed improvements to the
- Dataset constructor APIs enhanced

* Version 1.2.1
- Same as 1.2.0; first version on PyPI

* Version 1.2.0
- Support for the new BGZF-based bam.pbi file format
- Support for PacBio BAM spec version 3.0b7; previous versions are
  rejected

* Version 1.1.0
- Added dataset functionality to pbcore.io
- Added dataset.py tools for CLI manipulation of DataSet XML files

* Version 1.0.0
- BAM fixes
- Better FOFN handling
- Recognize additional part numbers for P6 chemistry
- Better --debug in ToolRunner

* Version 0.9.5
- BAM fixes
- Adopt aStart, aEnd conventions from BAM spec, replacing rStart, rEnd
- AlnIndex sharing among CmpH5Reader objects

* Version 0.9.4
- BAM support moved to 3.0 spec; support for earlier PacBio BAMs
  dropped
- Deprecation warning decorators added
- BAM-incompatible cmp.h5 accessors deprecated
- Moved to support 3.0 FASTA conventions in Fasta readers
- Rename FastaTable to IndexedFastaReader

* Version 0.9.3
- ".open" script added for convenience
- openers added (factory methods invoking the appropriate Reader
  class; useful for applications that want to transparently use either
  BAM or cmp.h5)

* Version 0.9.2
- BAM support: Addition of BamReader, IndexedBamReader, and BamAlignment
- Minor CmpH5Reader API changes for greater compatibility with
  BamReader (deprecation of movieInfo in favor of readGroupInfo)
- Removed unused components from CmpH5Reader API
- Add example BAM file
- Length accessors for FAST[AQ] record types

* Version 0.9.1 (SMRTanalysis 2.3.0p1)
- FASTA header parsing into "id" and "metadata" now available in the
  FastaRecord types

* Version 0.9.0 (SMRTanalysis 2.3.0)
- pbcore.chemistry: a new subpackage for decoding barcode information
  to the human-readable chemistry name
- BasH5Reader: more robust handling of broken region tables
- CmpH5Reader: loading an empty cmp.h5 will raise an EmptyCmpH5Error.
  This is because the semantics of an empty cmp.h5 were never defined,
  and for example it is not defined whether or not a cmp.h5 lacking a
  movie table is compliant.
